feat: возможность создания review request-а

alinkalina opened the pull request:
**по ТЗ:**
- routers\p2p_request.py - добавлен get-запрос, возвращающий ссылку на репозиторий и комментарий в словаре
- sql\crud.py - в класс P2PRequestCrud добавлен метод start_review, в котором реализован поиск самого старого ожидающего проверки проекта и назначается reviewer, статус проекта переводится в progress

__задачки со *:__
- sql\models.py и sql\schemas.py - добавлено поле review_start_date, в которое в методе start_review записывается время начала ревью

**+ по моей инициативе :) :**
- если в базе нет pending-проектов, запрос возвращает соответствующее сообщение вместо ссылки и комментария
- пользователю не может быть выдана ссылка на его же проект, для этого добавлен фильтр в start_review

alinkalina committed:
feat: возможность создания review request-а

**по ТЗ:**
- routers\p2p_request.py - добавлен get-запрос, возвращающий ссылку на репозиторий и комментарий в словаре
- sql\crud.py - в класс P2PRequestCrud добавлен метод start_review, в котором реализован поиск самого старого ожидающего проверки проекта и назначается reviewer, статус проекта переводится в progress

**задачки со *:**
- sql\models.py и sql\schemas.py - добавлено поле review_start_date, в которое в методе start_review записывается время начала ревью

**+ по моей инициативе :) :**
- если в базе нет pending-проектов, запрос возвращает соответствующее сообщение вместо ссылки и комментария
- пользователю не может быть выдана ссылка на его же проект, для этого добавлен фильтр в start_review

Roman821 reviewed (dismissed):
Поставленная задача выполнена! Собственная инициатива - круто!
Just одобряю!
Огромный молодчик!
Круто-круто-круто!

LaGGgggg commented on API/sql/crud.py:
```diff
@@ -64,3 +66,17 @@ def get_user_scopes(self, user: schemas.User) -> list[models.Scope]:
+        project = self.db.query(self.model) \
+            .filter(self.model.review_state == 'pending', self.model.creator_id != reviewer.id) \
+            .order_by(self.model.publication_date) \
+            .first()
```
Судя по всему, это сделано на основе рекомендаций по написанию запросов SQL, но для python, на мой взгляд, это менее читабельный вариант, чем, например, так:
```python
project = self.db.query(self.model).filter(
            self.model.review_state == 'pending', self.model.creator_id != reviewer.id
).order_by(self.model.publication_date).first()
```

В [PEP-8](https://peps.python.org/pep-0008/#maximum-line-length) это даже явно прописано:
> The preferred way of wrapping long lines is by using Python’s implied line continuation inside parentheses, brackets and braces. Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line continuation.<br>
> Перевод:
Предпочтительный способ переноса длинных строк — использование подразумеваемого продолжения строки Python внутри скобок и фигурных скобок. Длинные строки можно разбить на несколько строк, заключив выражения в круглые скобки. Их следует использовать вместо обратной косой черты для продолжения строки.

И ещё упоминание этой рекомендации:
> Continuation lines should align wrapped elements either vertically using Python’s implicit line joining inside parentheses, brackets and braces, or using a hanging indent [[1]](https://peps.python.org/pep-0008/#fn-hi). When using a hanging indent the following should be considered; there should be no arguments on the first line and further indentation should be used to clearly distinguish itself as a continuation line:<br>
> Перевод:
Строки продолжения должны выравнивать обернутые элементы либо по вертикали, используя неявное соединение линий Python внутри круглых скобок, скобок и фигурных скобок, либо с помощью висячего отступа [1]. При использовании висячего отступа следует учитывать следующее;
в первой строке не должно быть аргументов, а дальнейший отступ должен использоваться, чтобы четко отличить ее от строки продолжения:

LaGGgggg commented on API/sql/crud.py:
```diff
@@ -8,6 +8,8 @@
 from abc import ABC
 
+from datetime import datetime
+
```
Тут в целом что-то страшное из import-ми случилось. Должно быть три группы: базовая библиотека, сторонние библиотеки, код проекта.
И вот ссылка на [PEP-8](https://peps.python.org/pep-0008/#imports):
> Imports should be grouped in the following order:
> 1. Standard library imports.
> 2. Related third party imports.
> 3. Local application/library specific imports.

LaGGgggg commented on API/routers/p2p_request.py:
```diff
@@ -24,3 +24,13 @@ def create_p2p_request(
+) -> dict:
+    repository_link, comment = crud.P2PRequestCrud(db).start_review(current_user)
+    if not repository_link and not comment:
+        return {'problem': 'There are not any pending projects'}
```
Может и придираюсь, но более широко используется в подобном контексте `error`, а не `problem`. К тому же, если привязываться к семантике английского языка, то это скорее ошибка, чем проблема, которую следует решить

LaGGgggg commented on API/routers/p2p_request.py:
```diff
@@ -24,3 +24,13 @@ def create_p2p_request(
+def p2p_request_review(
+        current_user: models.User = Security(login_manager, scopes=['p2p_request']),
+        db: Session = Depends(get_db)
+) -> dict:
```
Лучше работать со схемами. По сути это те же словари, тот же json, но придуманы специально для типизации ответа API. Нужны для того, чтобы мы могли **точно** знать что вернёт API. Сейчас, согласно типизации, нам может вернуться **любой** json. Не очень безопасно и удобно (как минимум, чтобы узнать что вернёт API, нужно сделать запрос)

LaGGgggg commented on API/sql/crud.py:
```diff
@@ -64,3 +66,17 @@ def get_user_scopes(self, user: schemas.User) -> list[models.Scope]:
+        project.review_start_date = datetime.now()
+        self.db.add(project)
+        self.db.commit()
+        return project.repository_link, project.comment
```
Мне кажется, что лучше возвращать объект из бд (или None, если его нет), чем две строки. Это увеличивает гибкость метода и кода в целом, улучшает безопасность (с этим проще работать, чем с двумя строками), плюс мы таким образом исключаем из crud логику, которая связана скорее с route, чем с crud

LaGGgggg commented on API/sql/crud.py:
```diff
@@ -64,3 +66,17 @@ def get_user_scopes(self, user: schemas.User) -> list[models.Scope]:
     def __init__(self, db: Session) -> None:
         super().__init__(models.P2PRequest, schemas.P2PRequestCreate, db)
+
+    def start_review(self, reviewer):
```
А почему на вход `reviewer`? Мы же используем только его id, зачем нам лишняя информация, это без нужды усложняет логику

LaGGgggg commented on API/sql/crud.py:
```diff
@@ -64,3 +66,17 @@ def get_user_scopes(self, user: schemas.User) -> list[models.Scope]:
+        project.reviewer_id = reviewer.id
+        project.review_state = 'progress'
+        project.review_start_date = datetime.now()
+        self.db.add(project)
```
`self.db.add(project)` не имеет смысла в данном контексте. При изменении атрибутов объекта бд и последующем `self.db.commit()` этого объекта, все изменения в его атрибутах занесутся в базу данных. Также для обновления данных следует использовать `update`, а не `add`

LaGGgggg commented on API/sql/crud.py:
```diff
@@ -64,3 +66,17 @@ def get_user_scopes(self, user: schemas.User) -> list[models.Scope]:
+        if not project:
+            return '', ''
+        project.reviewer_id = reviewer.id
+        project.review_state = 'progress'
```
Лучше не "хардкодить" такие значения. Если в моделе это поле поменяется, то придётся искать все места, где оно встречается и менять его там (а можно и не найти). Лучше в [models.py](https://github.com/LaGGgggg/yp_p2p_api/blob/fcc321562b759b2b365c17d26a5765aeb166a51a/API/sql/models.py#L43) вынести в константу (в теле класса) этот "pending" и брать это значение уже из модели

LaGGgggg reviewed (changes requested):
Всё работает, это очень хорошо, работает стабильно. Что улучшить есть, это считай всегда так. Что заметил, написал.
Ещё момент по коду в целом: мне тяжеловато читать некоторые места, так как код не побит на логические блоки пустыми строками. Это не преступление, так точно можно, но лучше бы делить, так будет проще глядеть на всё это. (цитата из [PEP-8](https://peps.python.org/pep-0008/#blank-lines): "Extra blank lines may be used (sparingly) to separate groups of related functions...")
Overall everything is good!

LaGGgggg commented:
И я тестами в ближайшее время займусь для этой функциональности (возможно в другой ветке уже будут, или в другом pull request-е)

alinkalina commented:
Ясненько, поправим)
Уточню, получается схема, которую нужно возвращать, это schemas.P2PRequest?

LaGGgggg commented:
Да

alinkalina committed:
fix: доработки и улучшения

Улучшены некоторые неточные места, а также вид кода

alinkalina commented:
Исправила всё (вроде бы), что было указано в [ревью](https://github.com/LaGGgggg/yp_p2p_api/pull/12/files/836037fe0e7b21b13afd91f274e7d35867546e5c), надеюсь правильно поняла про константы в модели...

LaGGgggg committed:
fix: небольшие улучшения в crud

Изменён порядок импортов в соответствии с PEP-8.
Из BaseCrud убран атрибут scheme (не используется).
В P2PRequestCrud/start_review добавлена типизация.

LaGGgggg committed:
fix: улучшена безопасность вариантов P2PRequest/review_state

sql: добавлен модуль models_choices_enums.py (для типизации полей моделей посредством Enum).
sql/models.py/P2PRequest: улучшена безопасность путём получения вариантов значений для review_state из ReviewStateChoicesEnum.
core/schemas.py: переделана логика на использование ReviewStateChoicesEnum вместо Literal.

LaGGgggg committed:
feat: добавлено возвращение схем из p2p_request_review

core/schemas.py:
  Добавлена схема ErrorResponse.
  В P2PRequest удалены поля creator и reviewer.

routers/p2p_request.py/p2p_request_review:
  Добавлена типизация.
  Ошибка отсутствия проекта для review теперь возвращается в схеме (ErrorResponse).
  Route теперь возвращает схему P2PRequest.

LaGGgggg committed:
feat: добавлена проверка на отсутствие активных review пользователя

Теперь пользователь не сможет начать review, если у него уже есть активное (возвращается ErrorResponse).

LaGGgggg commented:
Добавил некоторые полезные штуки (и пофиксил недоделки небольшие)

LaGGgggg reviewed (approved):
В целом всё готово к merge-у, всё хорошо

Roman821 reviewed (approved):
Всё вроде крто! 👍
